group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1056,7 +1056,12 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
 
         // Get or create the consumer group.
         boolean createIfNotExists = memberEpoch == 0;
-        final ConsumerGroup group = getOrMaybeCreateConsumerGroup(groupId, createIfNotExists);
+        ConsumerGroup group;
+        if (maybeDeleteEmptyClassicGroup(groupId, records)) {
+            group = new ConsumerGroup(snapshotRegistry, groupId, metrics);
+        } else {
+            group = getOrMaybeCreateConsumerGroup(groupId, createIfNotExists);
+        }
```

Comment:
The method is already large so I wonder if we could push this logic into `getOrMaybeCreateConsumerGroup` and we should actually do this only if `createIfNotExists` is `true`.